### ISSUE TITLE
Way to set server response headers depending on client

### ISSUE THREAD

**Issue by Ossada** [aaugustin/websockets#617]
I am writing a general [OCPP](https://www.openchargealliance.org/protocols/ocpp-20/) server, that should support both ocpp 1.6 and ocpp 2.0. The protocol version is passed in  **Sec-WebSocket-Protocol** header from the client, where multiple versions are possible. Server should pick one and send the chosen version in response in the same header. I managed to hardcode response with _extra_headers=_ keyword like so:
```python    
start_server = websockets.serve(
        server.run, const.ADDR, const.PORT, process_request=check_endpoint,
        extra_headers={'Sec-WebSocket-Protocol': 'ocpp2.0'})
```
but this (obviously) works only for clients that support ocpp2.0. I searched for clues in issues and found this one #373 that proccesses pasword and username before establishing connection (if I understand this correctly). Is there a way to modify this class to set response protocol header?
Something like this:
```python 
class CheckProtocol(websockets.WebSocketServerProtocol):
    async def process_request(self, path, request_headers):
        protocol = request_headers["Sec-WebSocket-Protocol"]
        if protocol == 'ocpp2.0':
            websocket.extra_headers = {'Sec-WebSocket-Protocol': 'ocpp2.0'}
        elif protocol == 'ocpp1.6':
            websocket.extra_headers = {'Sec-WebSocket-Protocol': 'ocpp1.0'}
```
```python
if __name__ == '__main__':
    server = Server()
    start_server = websockets.serve(
        server.run, const.ADDR, const.PORT, process_request=check_endpoint,
        create_protocol=CheckProtocol)
```
This doesn't work, client doesn't get a correct response and disconnects. Is this the right way, or is there a much simpler way to set response headers? 
Thank you for your help.

**Comment by aaugustin** [comment]
Can you solve this by passing a callable `extra_headers`?

From the docs (emphasis added):

> extra_headers sets additional HTTP response headers — it can be a Headers instance, a Mapping, an iterable of (name, value) pairs, **or a callable taking the request path and headers in arguments and returning one of the above**

```python

def extra_headers(path, request_headers)```:
        protocol = request_headers.get('Sec-WebSocket-Protocol')
        if protocol == 'ocpp2.0':
            return {'Sec-WebSocket-Protocol': 'ocpp2.0'}
        elif protocol == 'ocpp1.6':
            return {'Sec-WebSocket-Protocol': 'ocpp1.0'}

start_server = websockets.serve(
        server.run, const.ADDR, const.PORT, process_request=check_endpoint,
        extra_headers=extra_headers)
```

**Comment by aaugustin** [comment]
(In the code I copy-pasted, I noticed that a request for version 1.6 is answered with version 1.0. Is that expected or a typo?)

**Comment by Ossada** [comment]
Thank you so much, it works! Yes, the ocpp1.0 thing was a typo, it should be ocpp1.6. Tested with multiple clients, I got some error:
Traceback (most recent call last):
  File "/home/vids/.local/lib/python3.6/site-packages/websockets/server.py", line 117, in handler
    extra_headers=self.extra_headers,
  File "/home/vids/.local/lib/python3.6/site-packages/websockets/server.py", line 514, in handshake
    for name, value in extra_headers:
TypeError: 'NoneType' object is not iterable
but it was my fault for not parsing the string (header can look like this: 'ocpp2.0, ocpp1.5' ) and of course extra_headers function returned None. 
When I fixed that, everything worked, thank you again!
